lib/kube/proxy/forwarder.go

```diff
@@ -732,6 +741,158 @@ func (f *Forwarder) newStreamer(ctx *authContext) (events.Streamer, error) {
 	return events.NewTeeStreamer(fileStreamer, f.cfg.StreamEmitter), nil
 }
 
+// join joins an existing session over a websocket connection
+func (f *Forwarder) join(ctx *authContext, w http.ResponseWriter, req *http.Request, p httprouter.Params) (resp interface{}, err error) {
+	f.log.Debugf("Join %v.", req.URL.String())
+
+	sess, err := f.newClusterSession(*ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := f.setupForwardingHeaders(sess, req); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if sess.noAuditEvents {
+		return f.remoteJoin(ctx, w, req, p, sess)
+	}
+
+	sessionIDString := p.ByName("session")
+	sessionID, err := uuid.Parse(sessionIDString)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	session := f.sessions[sessionID]
+	if session == nil {
+		return nil, trace.NotFound("session %v not found", sessionID)
+	}
+
+	if !session.tty {
+		return nil, trace.NotFound("session %v is not interactive", sessionID)
+	}
+
+	ws, err := f.upgrader.Upgrade(w, req, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	stream, err := streamproto.NewSessionStream(ws, streamproto.ServerHandshake{MFARequired: session.PresenceEnabled})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	client := &websocketClientStreams{stream}
+	party := newParty(*ctx, stream.Mode, client)
+	go func() {
+		<-stream.Done()
+		session.mu.Lock()
+		defer session.mu.Unlock()
+		session.leave(party.ID)
+	}()
+
+	err = session.join(party)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	<-party.closeC
+
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return nil, nil
+}
+
+// remoteJoin forwards a join request to a remote cluster.
+func (f *Forwarder) remoteJoin(ctx *authContext, w http.ResponseWriter, req *http.Request, p httprouter.Params, sess *clusterSession) (resp interface{}, err error) {
+	dialer := &websocket.Dialer{
+		TLSClientConfig: sess.tlsConfig,
+		NetDialContext:  sess.DialWithContext,
+	}
+
+	url := "wss://" + req.URL.Host
+	if req.URL.Port() != "" {
+		url = url + ":" + req.URL.Port()
+	}
+	url = url + req.URL.Path
+
+	wsTarget, respTarget, err := dialer.Dial(url, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer wsTarget.Close()
+	defer respTarget.Body.Close()
+
+	wsSource, err := f.upgrader.Upgrade(w, req, nil)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer wsTarget.Close()
+
+	err = wsProxy(wsSource, wsTarget)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return nil, nil
+}
+
+// wsProxy proxies a websocket connection between two clusters transparently to allow for
+// remote joins.
+func wsProxy(wsSource *websocket.Conn, wsTarget *websocket.Conn) error {
```

Comment:
We have something very similar to this in desktop access, and I just cleaned it up to make sure that it doesn't print a bunch of EOF errors when someone terminates a session.

Just pointing this out in case something similar might apply here:

https://github.com/gravitational/teleport/pull/10232